packages/@lwc/template-compiler/src/__tests__/fixtures/directive-lwc-if-else/dynamic/if-elseif-else/expected.js

```diff
@@ -0,0 +1,27 @@
+import { parseFragment, registerTemplate } from "lwc";
+const $fragment1 = parseFragment`<h1${3}>if condition</h1>`;
+const $fragment2 = parseFragment`<h1${3}>elseif condition</h1>`;
+const stc0 = {
+  key: 5,
+};
+function tmpl($api, $cmp, $slotset, $ctx) {
+  const {
+    st: api_static_fragment,
+    fr: api_fragment,
+    dc: api_dynamic_component,
+  } = $api;
+  return [
+    $cmp.visible.if
```

Comment:
For every branch of the `if` directive, we always generate a fragment with the same key, and each of those fragments needs to be diffed dynamically (because two branches need to be diffed when changing branches).

However, some of the branches do not have dynamic content, and in theory, they might be diffed statically (when diffing the same branch). For example, given this template:

```html
<template>
    {msg}
    <template lwc:if={visible}>
        <div>if condition</div>
        <div>was set to true</div>
    </template>
    <template lwc:else>
        <p>else condition</p>
    </template>
</template>
```

when `msg` changes, the diffing algo will patch the same branch using updateDynamicChildren (fragments are unstable), but in theory it can be patched using the updateStatichChildren routine.

What about:

- Always generate a fragment (unstable) for the if directive (the whole decision tree).
- For each branch, of the if, generate a fragment (stable or not will depend on the children), but with a different key.

for this example, this will be the new output:

```js
return [
    api_fragment(
        0,
        [
            $cmp.visible.if
                ? api_fragment(1, [api_static_fragment($fragment1(), 2)], 1 /* stable */)
                : $cmp.visible.elseif
                    ? api_fragment(2, [api_static_fragment($fragment2(), 4)], 1 /* stable */)
                    : api_fragment(
                        3,
                        [api_dynamic_component("x-foo", $cmp.trackedProp.foo, stc0)],
                        1 /* unstable */
                    ),
        ],
        0, // unstable.
    ),
];
```

in this way, there's always going to by a dynamic diffing of 1 element for the if (the 2 branches), and the branch itself going to be patched with the optimal routine.

Reply:
So, it would be okay if the two vNodes have different keys if they're all inside a vNode with the same key? (I thought if we were replacing the vnode, it _had_ to have the same key)

For your example, we would be replacing `api_fragment(2, ...)` with `api_fragment(3, ...)` when `$cmp.visible.elseif` changes. You're saying that is okay, since the entire thing is wrapped inside an `unstable` fragment so it will go through the dynamic diffing?

If so, I like this idea a lot because it'll let us have some performance optimization from knowing when the branches are going to be static. Much better than what we're doing now. I'll try it out this afternoon.

Reply:
> So, it would be okay if the two vNodes have different keys if they're all inside a vNode with the same key? (I thought if we were replacing the vnode, it _had_ to have the same key)

I don't fully understand the question. if 2 vnodes have the same key, they are going to be patched, which is different than replaced (mount + remove)


> For your example, we would be replacing `api_fragment(2, ...)` with `api_fragment(3, ...)` when `$cmp.visible.elseif` changes. You're saying that is okay, since the entire thing is wrapped inside an `unstable` fragment so it will go through the dynamic diffing?

yes, for the wrapping fragment (fr 0) always need to be unstable, so it uses the dynamic algo. remember that updateStaticChildren does not know how to "replace".

Reply:
@jodarove I tried it out, and although I like this as an optimization, it does require more extensive changes and possibly refactoring I'd like to do in a separate PR.

Right now, our `shouldFlatten` logic only checks the immediate children of the node. For a proper `isStable` check, we need to check the full tree of children until we see something that marks the tree as unstable (e.g. a nested for-block X levels down). We could just do that traversal when we transform the IfBlock, but I'd like to explore if there's a better way.

Right now, this PR is getting pretty big, and we're having to resolve merge conflicts with master every single week. Is it alright with you if we explore this optimization as a separate PR?

Reply:
sure, it makes sense. let's see it in another PR.